ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/builders/BundleErrorReporter.java

```diff
@@ -1232,7 +1232,39 @@ private void validateImportPackage(IProgressMonitor monitor) {
 							addMarkerAttribute(marker,PDEMarkerFactory.compilerKey,CompilerFlags.P_UNRESOLVED_IMPORTS);
 							return;
 						}
-					} else {
+
+						Map<String, Object> exportAttributes = export.getAttributes();
+						Map<String, Object> exportDirectives = export.getDirectives();
+						if (exportDirectives != null) {
+							if (exportDirectives.containsKey(Constants.RESOLUTION_MANDATORY)) {
+								Map<String, Object> importAttributes = importSpec.getAttributes();
+								String[] mandatories = (String[]) exportDirectives.get(Constants.RESOLUTION_MANDATORY);
+								for (String mandatory : mandatories) {
+									if (importAttributes == null || !importAttributes.containsKey(mandatory)
+											|| importAttributes.get(mandatory).toString() != exportAttributes
```

Comment:
You should (almost) never compare strings by their reference, since it is easy to have different string object with the same content.